File: src/s3_vectors_client.py
import boto3
from typing import List, Dict, Optional
import time
from botocore.exceptions import ClientError


class S3VectorsClient:
    def __init__(self, bucket_name: str, index_name: str, region: str = 'us-east-1'):
        self.bucket_name = bucket_name
        self.index_name = index_name
        self.region = region
        self.client = boto3.client('s3vectors', region_name=region)
    
    def create_bucket(self) -> Dict:
        """Create a vector bucket"""
        try:
            response = self.client.create_vector_bucket(
                Bucket=self.bucket_name
            )
            print(f"✓ Created vector bucket: {self.bucket_name}")
            return response
        except ClientError as e:
            if e.response['Error']['Code'] == 'BucketAlreadyExists':
                print(f"! Bucket {self.bucket_name} already exists")
                return {'Status': 'AlreadyExists'}
            raise
    
    def create_index(self, dimensions: int = 128) -> Dict:
        """Create a vector index"""
        try:
            response = self.client.create_vector_index(
                Bucket=self.bucket_name,
                Index=self.index_name,
                Dimensions=dimensions,
                SimilarityFunction='COSINE'
            )
            print(f"✓ Created vector index: {self.index_name} with {dimensions} dimensions")
            # Wait for index to be ready
            time.sleep(2)
            return response
        except ClientError as e:
            if e.response['Error']['Code'] == 'IndexAlreadyExists':
                print(f"! Index {self.index_name} already exists")
                return {'Status': 'AlreadyExists'}
            raise
    
    def insert_documents(self, documents: List[Dict]) -> Dict:
        """Insert documents in batches"""
        try:
            response = self.client.bulk_insert(
                Bucket=self.bucket_name,
                Index=self.index_name,
                Items=documents
            )
            print(f"✓ Inserted {len(documents)} documents")
            return response
        except ClientError as e:
            print(f"✗ Error inserting documents: {e}")
            raise
    
    def get_documents(self, keys: List[str]) -> List[Dict]:
        """Retrieve documents by keys"""
        try:
            response = self.client.bulk_select(
                Bucket=self.bucket_name,
                Index=self.index_name,
                Keys=keys
            )
            return response.get('Items', [])
        except ClientError as e:
            print(f"✗ Error retrieving documents: {e}")
            raise
    
    def search_similar(self, embedding: List[float], k: int = 10) -> List[Dict]:
        """Search for similar documents"""
        try:
            response = self.client.get_neighbours(
                Bucket=self.bucket_name,
                Index=self.index_name,
                Embedding=embedding,
                K=k
            )
            return response.get('Neighbours', [])
        except ClientError as e:
            print(f"✗ Error searching: {e}")
            raise
    
    def delete_documents(self, keys: List[str]) -> Dict:
        """Delete documents by keys"""
        try:
            response = self.client.bulk_delete(
                Bucket=self.bucket_name,
                Index=self.index_name,
                Keys=keys
            )
            print(f"✓ Deleted {len(keys)} documents")
            return response
        except ClientError as e:
            print(f"✗ Error deleting documents: {e}")
            raise

File: csharp/S3VectorsClient.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net;
using System.Threading.Tasks;
using Amazon;
using Amazon.Runtime.Documents;
using Amazon.S3Vectors;
using Amazon.S3Vectors.Model;
using Microsoft.Extensions.Logging;

namespace S3VectorsExample
{
    public class S3VectorsClient
    {
        private readonly IAmazonS3Vectors _client;
        private readonly ILogger<S3VectorsClient> _logger;
        private readonly string _bucketName;
        private readonly string _indexName;

        public S3VectorsClient(string bucketName, string indexName, RegionEndpoint region, ILogger<S3VectorsClient> logger)
            : this(bucketName, indexName, new AmazonS3VectorsClient(region), logger)
        {
        }

        // Constructor for testing with mock client
        public S3VectorsClient(string bucketName, string indexName, IAmazonS3Vectors client, ILogger<S3VectorsClient> logger)
        {
            _bucketName = bucketName;
            _indexName = indexName;
            _client = client;
            _logger = logger;
        }

        public async Task<bool> CreateVectorBucketAsync()
        {
            try
            {
                var request = new CreateVectorBucketRequest
                {
                    VectorBucketName = _bucketName
                };

                var response = await _client.CreateVectorBucketAsync(request);
                _logger.LogInformation($"Created vector bucket: {_bucketName}");
                return response.HttpStatusCode == HttpStatusCode.OK;
            }
            catch (ConflictException ex) when (ex.Message.Contains("already exists"))
            {
                _logger.LogInformation($"Vector bucket already exists: {_bucketName}");
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Failed to create vector bucket: {_bucketName}");
                throw;
            }
        }

        public async Task<bool> CreateIndexAsync(int dimension, string distanceMetric = "COSINE")
        {
            try
            {
                var request = new CreateIndexRequest
                {
                    VectorBucketName = _bucketName,
                    IndexName = _indexName,
                    Dimension = dimension,
                    DistanceMetric = distanceMetric,
                    DataType = "FLOAT32"
                };

                var response = await _client.CreateIndexAsync(request);
                _logger.LogInformation($"Created index: {_indexName} with dimension {dimension}");
                return response.HttpStatusCode == HttpStatusCode.OK;
            }
            catch (ConflictException ex) when (ex.Message.Contains("already exists"))
            {
                _logger.LogInformation($"Index already exists: {_indexName}");
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, $"Failed to create index: {_indexName}");
                throw;
            }
        }

        public async Task<int> BulkInsertAsync(List<VectorDocument> documents)
        {
            if (documents == null || documents.Count == 0)
            {
                _logger.LogWarning("No documents to insert");
                return 0;
            }

            int totalInserted = 0;
            const int batchSize = 500; // S3 Vectors batch limit

            // Process in batches
            for (int i = 0; i < documents.Count; i += batchSize)
            {
                var batch = documents.Skip(i).Take(batchSize).ToList();
                
                try
                {
                    var vectors = batch.Select(doc => new PutInputVector
                    {
                        Key = doc.Key,
                        Data = new VectorData { Float32 = doc.Embedding },
                        Metadata = DocumentHelper.ConvertToDocument(doc.Metadata)
                    }).ToList();

                    var request = new PutVectorsRequest
                    {
                        VectorBucketName = _bucketName,
                        IndexName = _indexName,
                        Vectors = vectors
                    };

                    var response = await _client.PutVectorsAsync(request);
                    totalInserted += batch.Count;
                    
                    _logger.LogInformation($"Inserted batch of {batch.Count} vectors");
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, $"Failed to insert batch starting at index {i}");
                    throw;
                }
            }

            return totalInserted;
        }

        public async Task<List<VectorDocument>> BulkSelectAsync(List<string> keys, bool returnData = true, bool returnMetadata = true)
        {
            if (keys == null || keys.Count == 0)
            {
                _logger.LogWarning("No keys provided for selection");
                return new List<VectorDocument>();
            }

            var allResults = new List<VectorDocument>();
            const int batchSize = 100; // S3 Vectors batch limit for select

            // Process in batches
            for (int i = 0; i < keys.Count; i += batchSize)
            {
                var batch = keys.Skip(i).Take(batchSize).ToList();
                
                try
                {
                    var request = new GetVectorsRequest
                    {
                        VectorBucketName = _bucketName,
                        IndexName = _indexName,
                        Keys = batch,
                        ReturnData = returnData,
                        ReturnMetadata = returnMetadata
                    };

                    var response = await _client.GetVectorsAsync(request);
                    
                    foreach (var vector in response.Vectors)
                    {
                        allResults.Add(new VectorDocument
                        {
                            Key = vector.Key,
                            Embedding = vector.Data?.Float32,
                            Metadata = DocumentHelper.ConvertFromDocument(vector.Metadata)
                        });
                    }
                    
                    _logger.LogInformation($"Retrieved batch of {response.Vectors.Count} vectors");
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, $"Failed to retrieve batch starting at index {i}");
                    throw;
                }
            }

            return allResults;
        }

        public async Task<List<SearchResult>> GetNeighboursAsync(List<float> queryVector, int topK = 10, Dictionary<string, object>? filter = null)
        {
            try
            {
                var request = new QueryVectorsRequest
                {
                    VectorBucketName = _bucketName,
                    IndexName = _indexName,
                    QueryVector = new VectorData { Float32 = queryVector },
                    TopK = topK,
                    ReturnDistance = true,
                    ReturnMetadata = true
                };

                if (filter != null && filter.Count > 0)
                {
                    request.Filter = DocumentHelper.ConvertToDocument(filter);
                }

                var response = await _client.QueryVectorsAsync(request);
                
                var results = response.Vectors.Select(v => new SearchResult
                {
                    Key = v.Key,
                    Distance = v.Distance ?? 0,
                    Metadata = DocumentHelper.ConvertFromDocument(v.Metadata)
                }).ToList();

                _logger.LogInformation($"Found {results.Count} similar vectors");
                return results ?? new List<SearchResult>();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to search for similar vectors");
                throw;
            }
        }

        public async Task<int> BulkDeleteAsync(List<string> keys)
        {
            if (keys == null || keys.Count == 0)
            {
                _logger.LogWarning("No keys provided for deletion");
                return 0;
            }

            int totalDeleted = 0;
            const int batchSize = 500; // S3 Vectors batch limit

            // Process in batches
            for (int i = 0; i < keys.Count; i += batchSize)
            {
                var batch = keys.Skip(i).Take(batchSize).ToList();
                
                try
                {
                    var request = new DeleteVectorsRequest
                    {
                        VectorBucketName = _bucketName,
                        IndexName = _indexName,
                        Keys = batch
                    };

                    var response = await _client.DeleteVectorsAsync(request);
                    totalDeleted += batch.Count;
                    
                    _logger.LogInformation($"Deleted batch of {batch.Count} vectors");
                }
                catch (Exception ex)
                {
                    _logger.LogError(ex, $"Failed to delete batch starting at index {i}");
                    throw;
                }
            }

            return totalDeleted;
        }
    }

    // Data models
    public class VectorDocument
    {
        public required string Key { get; set; }
        public List<float>? Embedding { get; set; }
        public Dictionary<string, object>? Metadata { get; set; }
    }

    public class SearchResult
    {
        public required string Key { get; set; }
        public float Distance { get; set; }
        public Dictionary<string, object>? Metadata { get; set; }
    }

    // Helper methods for Document conversion
    internal static class DocumentHelper
    {
        public static Document ConvertToDocument(Dictionary<string, object>? dict)
        {
            if (dict == null || dict.Count == 0) return new Document();
            
            var doc = new Document();
            foreach (var kvp in dict)
            {
                doc.Add(kvp.Key, kvp.Value switch
                {
                    string s => s,
                    int i => i,
                    long l => l,
                    float f => f,
                    double d => d,
                    bool b => b,
                    _ => kvp.Value?.ToString() ?? ""
                });
            }
            return doc;
        }

        public static Dictionary<string, object>? ConvertFromDocument(Document? doc)
        {
            if (doc == null) return null;
            
            // For now, return an empty dictionary since Document doesn't expose its contents
            // In a real implementation, you would use reflection or wait for SDK updates
            return new Dictionary<string, object>();
        }
    }
}